### PR TITLE
cmd/swarm: Add pss smoke test stub

### PR DESCRIPTION
**Pull request by nolash**
This PR adds a modest beginning on how a pss smoke test could be set up.

It is intended as a stub for anyone who feels like it to elaborate on.

### CODE REVIEW

File: cmd/swarm/swarm-smoke/pss.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+		defer cancel()
+		sub, err := rpcClient.Subscribe(ctx, "pss", session.msgC, "receive", session.topic, true, false)
+		if err != nil {
+			log.Error("Error calling host for subscribe", "err", err)
+			continue
+		}
+		session.nodes = append(session.nodes,
+			&pssNode{
+				hostIdx: i,
+				//msgC:           msgC,
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: remove?

File: cmd/swarm/swarm-smoke/pss.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+// keyed by hex representation of overlay address
+type pssSession struct {
+	topic        pss.Topic
+	msgC         chan pss.APIMsg
+	nodes        []*pssNode
+	requiredJobs map[string]*pssJob
+	allowedJobs  map[string]*pssJob
+}
+
+func pssChecks(ctx *cli.Context, tuid string) error {
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: not sure if this is necessary. you can just wrap `pssAsym` directly

File: cmd/swarm/swarm-smoke/pss.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+	return pssAsym(ctx, tuid)
+}
+
+func pssAsym(ctx *cli.Context, tuid string) error {
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: can we please have a brief text describing what we are testing, what are the test's erroneous stop conditions (if they at all exist), does the test have a green exit condition? or does it just go on running till it fails?

File: cmd/swarm/swarm-smoke/pss.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+		err = rpcClient.Call(&pubkey, "pss_getPublicKey")
+		if err != nil {
+			log.Error("Error calling host for pubkey", "err", err)
+			continue
+		}
+
+		// subscribe to the topic for the session
+		// this creates the incoming message handler automatically
+		// any message received with this topic comes on the channel
+		//msgC := make(chan pss.APIMsg)
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: remove?

File: cmd/swarm/swarm-smoke/pss.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+		defer cancel()
+		sub, err := rpcClient.Subscribe(ctx, "pss", session.msgC, "receive", session.topic, true, false)
+		if err != nil {
+			log.Error("Error calling host for subscribe", "err", err)
+			continue
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: should we continue or fail the whole test?
what about metrics? I'd like to see a counter being incremented here

File: cmd/swarm/swarm-smoke/pss.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	recvNodeIdx := rand.Intn(len(session.nodes)-1) + 1
+	recvNode := session.nodes[recvNodeIdx]
+
+	// set recipient in pivot node
+	err := senderNode.client.Call(nil, "pss_setPeerPublicKey", recvNode.pubkey, session.topic, "0x")
+	if err != nil {
+		return err
+	}
+
+	// for asym it's not necessary to set recipient, but for sym we must
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: remove code block?

File: cmd/swarm/swarm-smoke/pss.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return err
+	}
+
+	// receive the msg or timeout and fail
+	// and check its validity
+	pctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	select {
+	//case res := <-(recvNode.msgC):
```
Comment (acud) [inline cmd/swarm/swarm-smoke/pss.go]: rm

**Review by acud — CHANGES_REQUESTED** [review]

File: cmd/swarm/swarm-smoke/pss.go (hunk 5 above)
Comment (skylenet) [reply cmd/swarm/swarm-smoke/pss.go]: I've update the test to send multiple messages (configurable via CLI flag `pss-messages`) via PSS using asymetric encryption between random nodes.

I've added the following counters:

`pss.asym.failMsg` - total failed messages
`pss.asym.successMsg` - total successful messages

File: cmd/swarm/swarm-smoke/main.go [hunk 8]
```diff
@@ -154,6 +161,12 @@ func main() {
 			Usage:   "measure network aggregate capacity",
 			Action:  wrapCliCommand("sliding-window", slidingWindowCmd),
 		},
+		{
+			Name:    "pss-asym",
+			Aliases: []string{"pa"},
+			Usage:   "PSS: send and receive multiple messages across random nodes using asymetric encryption",
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/main.go]: asymmetric

File: cmd/swarm/swarm-smoke/pss.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+// send messages using pss asymetric encryption across random nodes
+func pssAsymCheck(ctx *cli.Context, tuid string) error {
+	// use input seed if it has been set
+	if inputSeed != 0 {
+		seed = inputSeed
+	}
+	rand.Seed(int64(seed))
+	if pssMessageCount <= 0 {
+		pssMessageCount = 1
+		log.Warn("message count should be a positive number. Defaulting to %d", pssMessageCount)
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: This should be structured log line - odd number of arguments, for example:
`log.Warn(fmt.Sprintf("message count should be positive. defaulting to %v", pssMessageCount))` in order to have 1 argument.

File: cmd/swarm/swarm-smoke/pss.go [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+				pubkey:         pubkey,
+				deregisterFunc: sub.Unsubscribe,
+			},
+		)
+	}
+
+	return session
+}
+
+// sends a single PSS message between two random nodes using asymetric encryption
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: comments in `godoc` start with the name of the method described.
`// pssAsymDo sends a single PSS message ...`

File: cmd/swarm/swarm-smoke/pss.go [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+func pssAsymDo(ctx *cli.Context, session *pssSession, tuid string) error {
+
+	senderNodeIdx := rand.Intn(len(session.nodes) - 1)
+	senderNode := session.nodes[senderNodeIdx]
+	log.Trace("sender node", "pss_baseAddr", hexutil.Encode(senderNode.addr), "host", hosts[senderNodeIdx])
+
+	// receiving node has to be different than sender
+	var recvNodeIdx int
+	for i := 0; i < 20; i++ {
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: probably better to do a while loop, although 20 times should be enough :)

File: cmd/swarm/swarm-smoke/pss.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+		recvNodeIdx = rand.Intn(len(session.nodes) - 1)
+		if recvNodeIdx != senderNodeIdx {
+			break
+		}
+	}
+	if recvNodeIdx == senderNodeIdx {
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: if you add a while loop above, this is not needed.

File: cmd/swarm/swarm-smoke/pss.go [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+		resIdx := toMsgIdx(res.Msg)
+		resJob, ok := session.requiredJobs[resIdx]
+		if !ok {
+			return fmt.Errorf("corrupt message for job %s", resIdx)
+		}
+		if !bytes.Equal(res.Msg, resJob.msg) {
+			return fmt.Errorf("message mismatch. expected: %s got: %s", resJob.msg, res.Msg)
+		}
+		log.Info("got msg", "job", hexutil.Encode([]byte(resIdx)))
+	case <-pctx.Done():
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: you can just use `time.After(1*time.Second)` - it is simpler than context with a timeout and cancel function.

File: cmd/swarm/swarm-smoke/pss.go [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+	case res := <-(session.msgC):
+		resIdx := toMsgIdx(res.Msg)
+		resJob, ok := session.requiredJobs[resIdx]
+		if !ok {
+			return fmt.Errorf("corrupt message for job %s", resIdx)
+		}
+		if !bytes.Equal(res.Msg, resJob.msg) {
+			return fmt.Errorf("message mismatch. expected: %s got: %s", resJob.msg, res.Msg)
+		}
+		log.Info("got msg", "job", hexutil.Encode([]byte(resIdx)))
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke/pss.go]: maybe `log.Debug`

**Review by nonsense — APPROVED** [review]
lgtm provided this works as expected on k8s

**Comment by skylenet** [comment]
@justelad @nonsense thanks for the feedback. Let me know if there is anything else. Otherwise I'll open this on upstream.

Next steps will be done on separate PRs:
- Add grafana dashboards for these metrics
- Also use symmetrical encryption on the PSS smoke tests

**Comment by nonsense** [comment]
The `swarm-smoke` binary is starting to grow. We have `sync`, `pss` and `feed` tests in it, we should review it at some point and maybe think about splitting, before it gets too large so that we introduce bugs in it.

**Comment by skylenet** [comment]
> The `swarm-smoke` binary is starting to grow. We have `sync`, `pss` and `feed` tests in it, we should review it at some point and maybe think about splitting, before it gets too large so that we introduce bugs in it.

@nonsense I think this is a good idea. So I've split the `pss` tests already to a new binary.

File: cmd/swarm/swarm-smoke-pss/pss.go [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+
+	msgIdx := toMsgIdx(randomMsg)
+	s.jobs[msgIdx] = &j
+
+	log.Debug("generated job", "job", hexutil.Encode([]byte(msgIdx)), "sender", hosts[j.sender.hostIdx], "recv", hosts[j.receiver.hostIdx])
+
+	return j
+}
+
+// waitForJob blocks until a msg is received or a timeout is reached
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke-pss/pss.go]: waitForMsg

File: cmd/swarm/swarm-smoke-pss/pss.go [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+type pssSession struct {
+	topic pss.Topic
+	msgC  chan pss.APIMsg
+	nodes []*pssNode
+	jobs  map[string]*pssJob
+}
+
+func pssAsymCheck(ctx *cli.Context, tuid string) error {
```
Comment (nonsense) [inline cmd/swarm/swarm-smoke-pss/pss.go]: nitpick, but i introduced `tuid` as a `test uuid` in order to get all logs from a specific test. this is no longer needed, as i've noticed we can filter by the job id on elastic search, so `tuid` is completely redundant now. also it is not passed to the `runCheck` function.

feel free to remove or leave as is for now.

Comment (skylenet) [reply cmd/swarm/swarm-smoke-pss/pss.go]: I've removed it.

**Comment by skylenet** [comment]
Opened upstream https://github.com/ethereum/go-ethereum/pull/19338

**Pull request closed without merging**